### ISSUE TITLE
Autocomplete of imports prepends "src/" in certain circumstances

### ISSUE THREAD

**Issue by daboyuka** [go-lang-plugin-org/go-lang-idea-plugin#2512]
Plugin version: 0.10.1296
IDE: IntelliJ IDEA 15.0.5
Java: "Java(TM) SE Runtime Environment (build 1.8.0_72-b15)"
OS: Mac OS X El Capitan 10.11.3

When autocompleting Go imports under certain circumstances (see below), the autocompletion prepends "src/" to the package path. For example, if the package is located at `foo/bar` on my gopath, I type `"bar"` into the import block, and I hit Ctrl+Space, it will autocomplete to `"src/foo/bar"`.

This behavior is new since upgrading to version 0.10.1296. I believe the working version before the update was the immediately previous release, as I accept updates regularly, but if someone can direct me to an update log, I will be happy to verify this.

My project/filesystem configuration is as follows (simplified):

* I have a directory `/myproject/src/foo/bar/` that contains some Go source files
* My GOPATH contains the path `/myproject/`
* The IntelliJ project directory is `/myproject/`

If I autocomplete packages on my GOPATH that are _not_ under `/myproject` (including system libraries), the correct path is used. However, if I use autocomplete a package under `/myproject/src`, it prepends "src". For example,  autocompleting`"bar"` inserts `"src/foo/bar"` instead of `"foo/bar"`, even though `/myproject/src` is on the GOPATH.

As I mentioned, this behavior is new in my current plugin version. I believe the issue is that the project directory is now being implicitly added to the GOPATH at a higher priority/order than the explicitly-configured GOPATH, whereas before the existing GOPATH took precedence. I know that my project directory structure is not ideal/idomatic for Golang, but this is an existing project that would be difficult to reorganize.

Thank you in advance for any help on this issue!

**Comment by dlsniper** [comment]
Hi,

What values do you have in ` File | Settings | Languages and Frameworks | Go | Go libraries ` ?
Thanks.

**Comment by daboyuka** [comment]
@dlsniper here you go. I'm going to switch to (almost) real paths, as this will be too confusing otherwise:

Global libraries:
* `/Users/me/gopath`
* `/Users/me/Documents/company/myproject`
* `/Users/me/Documents/company/otherproject`
* `/usr/local/go_appengine/goroot`
Project and module libraries: none and none.
`Use GOPATH that's defined in system environment` is checked.

My IntelliJ project directory is `/Users/me/Documents/company/myproject`. Imports work fine for packages under `otherproject` but not `myproject`.

**Comment by dlsniper** [comment]
Hi,

I can't replicate the issue, please submit your ` .idea/libraries/GOPATH__<project>_.xml ` and ` <file>.iml` (not sure where that is, might be under the project root or `.idea` folder). Feel free to anonimize the paths but please make sure you keep the same rename for each path across different instances.

Also, please remove the `myproject` and `otherproject` project from the `Global Libraries` and use the `Project  Libraries` and set it to the `otherproject` for `myproject` (and vice versa). Remove the `go_appengine` as well as that is the Go SDK, you should use that one instead of a normal Go SDK.

Thanks.

**Comment by daboyuka** [comment]
Sure, I'll attach the files here as soon as I can, hopefully later today. Thanks for your help!

**Comment by daboyuka** [comment]
Here are the files you requested: [intellij-golang-issue2512.zip](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/files/256315/intellij-golang-issue2512.zip)


I believe I found the issue and a solution (or is it a workaround?). Under `File | Project Structure | Modules | Dependencies`, the problem occurs whenever the `<Module source>` entry is above the `whatever your Go SDK is called` entry, and goes away when the reverse is true. Changing the order of these entries has an effect after doing `File | Invalidate Caches / Restart`, `Invalidate and Restart` to force re-indexing (possibly only the invalidation is necessary, not the restart).

So, the following steps should reproduce:
1. Set up a project with project directory the same as some entry on the GOPATH (e.g., `/myproject` is both the project directory and a GOPATH path)
2. Go to `File | Project Structure | Modules | Dependencies` and move `<Module source>` above `whatever your Go SDK is called` and click `Apply`
3. Go to `File | Invalidate Caches / Restart` and click `Invalidate and Restart`
4. Import paths within the project directory should now incorrectly autocomplete with "src/" prefix
5. Repeat steps 2 and 3 but reverse the order of the two Dependency entries
6. Import paths within the project directory should now autocomplete correctly

**Comment by daboyuka** [comment]
Nope, the conditions for reproducing are more complex, it seems; I managed to break it again while experimenting and the steps above have no effect now. I'll keep tinkering and report any findings.

**Comment by daboyuka** [comment]
Well, after some more tinkering, it's back to working again, and nothing I do breaks it. The only final differences from the files linked in my earlier comment are:
1. I managed to remove the `<orderEntry type="module-library">` block from the `.iml` file by removing that explicit dependency from `File | Project Structure | Modules | Dependencies`; it was redundant with the library/GOPATH entry in `File | Settings | Languages and Frameworks | Go | Go libraries` for `otherproject`.
2. The `<root url="...">` entries in `.idea/libraries/GOPATH__<project>_.xml` are in a different (random) order.

I suspect there may be some lingering cached info hanging around despite `File | Invalidate Caches / Restart`.

To my limited understanding, the XML chunk `<excluded><root url="file://$PROJECT_DIR$" /></excluded>` in `.idea/libraries/GOPATH__<project>_.xml` looks suspicious, but it seems to have no impact on this issue whether it is present or absent (via manual file edit).

As long as this keeps working, I personally am happy. However, I'm also willing to keep digging into this by trying anything you may suggest, if you wish.

**Comment by daboyuka** [comment]
I suspect there may be some lingering cached info hanging around despite `File | Invalidate Caches / Restart`.

To my limited understanding, the XML chunk `<excluded><root url="file://$PROJECT_DIR$" /></excluded>` in `.idea/libraries/GOPATH__<project>_.xml` looks suspicious, but it seems to have no impact on this issue whether it is present or absent (via manual file edit).

As long as this keeps working, I personally am happy. However, I'm also willing to keep digging into this by trying anything you may suggest, if you wish.

**Comment by dlsniper** [comment]
Hi,

I think you've overdone the setup for the SDK / workspace :D Go doesn't require any changes into the LIbraries, or ordering on how things work in the configuration.

I'd suggest removing all the Go SDKs you have and manually configured external libraries as well as as everything from Go libraries and maybe even the .idea folders and .iml files from your projects.

The next step would be to invalidate the caches. close the current project and manually restart the IDE (don't choose the invalidate and restart option).

Once this is done, to setup a working environment please do the following:
- choose open folder to create a new project
- once the project is opened, open a Go file
- you should see two yellow prompts at the top of the editor:
 - one to setup the SDK
 - one to change the module to a Go module
- chose the setup SDK and create a new Go SDK (or a new Go AppEngine SDK, or both)
- then select one of the SDKs you've added and click OK
- if you see now two prompts about `GOPATH is empty` then follow that and configure the libraries as I've mentioned in the previous comment.
- finally choose to change the module type to Go and everything should work now

Please let me know if you have any issues with this.

**Comment by zolotov** [comment]
@daboyuka why do you need `/usr/local/go_appengine/goroot` in global libraries? Is it app-engine SDK? If so then it shoudn't be a part of GOPATH.

**Comment by daboyuka** [comment]
@dlsniper, I followed your instructions: removed all Go SDKs, deleted `*.iml` and `.idea/**`, invalidate/restart. I additionally reinstalled IntelliJ IDEA (somehow got the idea it might be out-of-date, but of course it wasn't) and the Go plugin. All of the prompts you mention in your instructions did indeed pop up, and I completed them all.

Unfortunately, the "src/" prepend problem is present even in this clean setup. My `.iml` and `GOPATH__myproject_.xml` files look substantially the same as before (sans some exclude paths and build flags, the latter of which are no longer relevant anyway). I can attach them again if you wish, though.

@zolotov you're right, it's not necessary. I configured my IDE back when I knew very little about IntelliJ or Go. It's gone from Global Libraries now, and only exists under the SDK itself as a source/class path.

**Comment by zolotov** [comment]
@daboyuka 

> I removed it from Global Libraries now; it only exists under the SDK itself as a source/class path.

So the problem with `src` still remains? If so, please create and attach an empty project that I can use for recreating the issue. Also please tell me what the version of GAE sdk you're using.

> I don't have the option for an AppEngine SDK;maybe that's only in the current dev version?

You don't need AppEngine SDK option, the plugin can figure out that your Go SDK is GAE SDK.

**Comment by daboyuka** [comment]
Sure, I will try to do that later today.

**Comment by daboyuka** [comment]
(Sorry, wrong button)

**Comment by daboyuka** [comment]
Sorry for the delay in getting back. I have tried making a new, empty project with the basic same directory structure and settings as my main project, but it does not seem to have the same issue, even after various combinations or restarts and cache invalidations. I've compared the `.idea/*` and `*.iml` files and see no qualitative differences. I'm not sure what is missing for a MWE.

My main project continues to do the `src/` thing only sometimes. Half the time it is persistent despite many restarts and/or `Invalidate Caches`, and the other half it works fine for one session immediately after an `Invalidate Caches` and manual restart then does the `src/` thing after another manual restart.

Additional finding: this issue affects not only autocompletion of imports, but also Find Usages: when the issue is occurring, Find Usages seems to not return results outside of the current package.

Is there any other information I can capture/send to help diagnose this issue?

**Comment by zolotov** [comment]
@daboyuka I don't know how to catch this issue. Have you done any manual changes in `Project Settings`? E.g. there is a strange module-library named `otherproject` in iml-file.

**Comment by daboyuka** [comment]
No, I performed the following steps on the main project (I tried to follow @dlsniper's suggested steps but I guess I got a couple of them out of order):
1. `Invalidate Caches` then manual exit
2. Delete `.idea/` and `*.iml`
3. Start IntelliJ, select `Open`, open the main project directory
4. Opened a `.go` file
5. Clicked the yellow notice about converting the project to a Go module + restart
6. Added my two libraries under `Preferences | Go | Go Libraries | Project libraries` (`myproject` and `otherproject`). Nothing else is on this page.
7. Clicked the yellow notice about selecting a Go SDK and selected my standard Go SDK (configured by selecting the Go App Engine main directory and letting IntellJ set up the configuration).
8. `Invalidate Caches` and manual restart just for good measure

The `.iml` looks qualitatively the same as the new, empty project with similar directory structure, but that project works fine.

One difference I noticed between the main project and empty project: autocomplete of imports doesn't work at all during `Indexing...` for the main project, but for the empty project it can autocomplete packages in the current project (though not outside the project, obviously). This seems to indicate that the empty project has its source files intrinsically indexed somehow even before the main indexing phase, whereas the main project does not for some reason.

**Comment by zolotov** [comment]
@daboyuka the iml that you attached here contains following library. I wonder how did you add it.

```xml
     <orderEntry type="module-library">
      <library name="otherproject">
        <CLASSES />
        <JAVADOC />
        <SOURCES>
          <root url="file://$MODULE_DIR$/../otherproject" />
        </SOURCES>
      </library>
    </orderEntry>
```

Please attach screenshot of all tabs of `File | Project Structure | Modules | <your module>` page.

**Comment by daboyuka** [comment]
The screenshots you requested are below. I give up anonymizing at this point, it's too much work :). Sorry for the confusion this may cause. The following equivalences apply:
`myproject` = `pendo-appengine`
`otherproject` = `goexternal`

My packages are under `src/pendo.io/<package>/*.go`. For example, if I open a file `src/pendo.io/types/types.go` and try to autocomplete an import `"types"`, it will autocomplete to `src/pendo.io/types` rather than `pendo.io/types`.

![screen shot 2016-05-23 at 13 10 05](https://cloud.githubusercontent.com/assets/3924734/15478399/ff7d405c-20e7-11e6-95b1-eacfedd085fc.png)
![screen shot 2016-05-23 at 13 10 32](https://cloud.githubusercontent.com/assets/3924734/15478398/ff7be16c-20e7-11e6-97b7-9fafe829635b.png)
![screen shot 2016-05-23 at 13 10 43](https://cloud.githubusercontent.com/assets/3924734/15478400/ff7edc82-20e7-11e6-8045-ef6b1e599a1f.png)

**Comment by daboyuka** [comment]
Here are fresh copies of my config files (non-anonymized), after having simplified my setup significantly.
[configs.zip](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/files/277918/configs.zip)

**Comment by zolotov** [comment]
The configuration looks good. I have an idea actually. Do you have any `yaml` file in `pendo-appengine` directory?

**Comment by daboyuka** [comment]
@zolotov Yes actually, three of them, for various non-IntellJ purposes. Could they be interfering somehow?

**Comment by zolotov** [comment]
@daboyuka yes, they can: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1525.

Looks like we should invent something smarter than this

**Comment by daboyuka** [comment]
@zolotov Oh, I also have some Google App Engine `.yaml` files under `modules/<module>/<module>.yaml`, So #1525 seems a likely cause.

Given this, the issue is probably that the parent directory `.../pendo-appengine` of the module directories `.../pendo-appengine/modules/<module>` is on the GOPATH.

In my specific instance, I don't need any special behavior from IntelliJ for handling GAE modules/paths. Can I just disable this functionality somehow?

**Comment by zolotov** [comment]
> In my specific instance, I don't need any special behavior from IntelliJ for handling GAE modules/paths. Can I just disable this functionality somehow?

Unfortunately not, the only way is to place yaml files outside project or module, e.g. by using `pendo.io` as a project root.

**Comment by daboyuka** [comment]
Oh well, that's fine. Based on your suggestion, I've reconfigured my project to have a typeless module at the top level and a Go module at `src/pendo.io`, and everything fine so far. I'll post back if I discover any further issue with this configuration. Thanks for sticking with me on this issue!
